### ISSUE TITLE
Refresh rate too slow - Moving to push mode

### ISSUE THREAD

**Issue by marcolivierarsenault** [marcolivierarsenault/moonraker-home-assistant#168]
Right now the integration refresh every 30 sec, which is too slow for many people. 

We will be pivoting this integration in pull mode instead of push.

**Temporary solution**
you can lower the refresh rate
https://github.com/marcolivierarsenault/moonraker-home-assistant/blob/ca5c81fcf24264f6bf3a7910cbb274829c15e9be/custom_components/moonraker/__init__.py#L29

**Comment by marcolivierarsenault** [comment]
no, tbh, no bandwidth has not progressed in months.

**Comment by theclue** [comment]
I use to create template switch that makes respectively an output pin from the board up or down; then, using value_template for a numeric sensor I get from the board to refresh the state of the switch.

Having 30 seconds makes the switch temporarly revert to the older state when commuting.

Hardcoding 2 seconds solved the issue brillantly, nonetheless having this pull-based looks definitively much more elegant :)

**Comment by jamincollins** [comment]
Why not simply change/update the interval when the printer becomes active?

**Comment by marcolivierarsenault** [comment]
yeah that makes sense. with the delay to the push mode, I will likely add a parameter to pull at a higher frequence

**Comment by marcolivierarsenault** [comment]
@jamincollins I have added an option to set a different timer. 

details here: https://moonraker-home-assistant.readthedocs.io/en/latest/connection.html#polling-rate-configuration


make sure to download latest (`main`) version of the code: https://moonraker-home-assistant.readthedocs.io/en/latest/support/latest.html

**Comment by jamincollins** [comment]
Thank you

**Comment by cf-mp** [comment]
Under configure tab in moonraker integrations on HA

> where is that optin for change pooling rate?

![image](https://github.com/user-attachments/assets/41b1b120-ed6c-4544-bd1b-f0632665ebba)

**Comment by adec** [comment]
Hi @marcolivierarsenault. Is there anything I can do to help move the integration toward push mode? Have you done any work or analysis on it that I can learn from?  I may be overthinking it but one of my concerns is whether we want to update all attributes in Home Assistant at the same frequency that they are pushed (although I've not done any analysis as yet on how often attributes are updated).

**Comment by marcolivierarsenault** [comment]
no sorry, no work done here.

**Comment by dewi-ny-je** [comment]
I don't expect it to work, but as starting point in case someone else has time to play with it I asked GPT 5.2 Thinking the following:

> this integration for Home Assistant connects to a Moonraker instance and periodically polls (pulls) updated entity values. Modify it to switch to a push model, where the integration subscribes to entity updates via Moonraker API, so that unnecessary updates for some non-changing entities are avoided, and more quickly changing entities can be tracked better. Introduce a configurable throttle parameter to avoid excessive updates, with default to 5 seconds.

and after 13 minutes (!) of thinking it came out with this answer:

> I modified the integration to use Moonraker’s **push / subscribe** model (WebSocket) instead of periodic polling for entity state, and added a **configurable throttling option** (default **5 seconds**) to limit how often HA is notified of incoming updates.
> 
> ### What changed
> 
> * **Switched from `printer.objects.query` polling → `printer.objects.subscribe` + `notify_status_update`**
> 
>   * Only changed fields are sent by Moonraker, so non-changing entities won’t be “updated” unnecessarily.
>   * Fast-changing objects (temps, motion, progress) can be tracked more responsively (bounded by throttle).
> * **Added option `push_throttle` (seconds, default 5)**
> 
>   * Push updates are merged immediately internally, but Home Assistant is notified at most once per throttle window.
>   * Set to `0` to effectively disable throttling (not recommended unless you really want very frequent updates).
> * **Printer state/message now come via subscription**
> 
>   * Added subscription to `webhooks.state` and `webhooks.state_message` and mirror them into `printer.info` so existing sensors keep working without polling `printer.info`.
> * **GCode metadata is fetched on-demand**
> 
>   * When the active filename changes in status updates, the integration pulls metadata once (debounced) instead of polling it repeatedly.
> * **Clean unload**
> 
>   * On unload, it now unregisters push callbacks and disconnects the websocket.
> 
> ### Where to configure throttle
> 
> Home Assistant UI → **Settings → Devices & Services → Moonraker → Configure**
> 
> * New field: **`push_throttle`** (seconds)
> 

and with this:
[custom_components_push.zip](https://github.com/user-attachments/files/24761285/custom_components_push.zip)

My expectations are low and I'm not able to check or test, but someone who actually knows how to code might.

Usually I would ask the same Gemini Thinking but it accepts only 10 files and the integration has 14, so I couldn't test to offer an alternative.
